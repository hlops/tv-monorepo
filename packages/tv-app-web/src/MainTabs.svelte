<script>
  import { WiredTab } from "./ui/wired/wired-tab";
  import { WiredTabs } from "./ui/wired/wired-tabs";
  import TvGuidesUpload from "./ui/tvGuides/TvGuidesUpload.svelte";
  import Channels from "./ui/channles/Channels.svelte";
  import TvGuides from "./ui/tvGuides/TvGuides.svelte";
</script>

<wired-tabs selected="Two">
  <wired-tab name="One" label="Каналы">
    <Channels />
  </wired-tab>
  <wired-tab name="Two" label="Телепрограмма">
    <TvGuides />
  </wired-tab>
  <wired-tab name="Three" label="Загрузка">
    <TvGuidesUpload />
  </wired-tab>
</wired-tabs>
