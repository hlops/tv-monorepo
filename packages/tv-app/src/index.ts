import fastify from "fastify";
import fastifyCors from "fastify-cors";
import fastifyMultipart from "fastify-multipart";
import * as fs from "fs";
import * as http from "http";
import { Readable } from "stream";
import playlistRoute from "./routes/playlist";
import tvGuideRoute from "./routes/tvGuide";

const server = fastify<http.Server, http.IncomingMessage, http.ServerResponse>({});

server.register(fastifyCors, { methods: ["GET", "PUT", "POST", "FETCH"] });

interface BodyEntry {
  data: Buffer;
  filename: string;
  encoding: string;
  mimetype: string;
  limit: false;
  tempFileName?: string;
}

server.register(fastifyMultipart, {
  addToBody: true,
  sharedSchemaId: "MultipartFileType",
  onFile: (
    fieldName: string,
    stream: Readable,
    filename: string,
    encoding: string,
    mimetype: string,
    body: Record<string, BodyEntry>
  ) => {
    console.log(fieldName, filename, encoding, mimetype);
    body.file.tempFileName = "./node_data/uploadFile";
    const ws = fs.createWriteStream(body.file.tempFileName);
    //    stream.resume();
    stream.pipe(ws);
  }
  /*
    limits: {
      fieldNameSize: 100, // Max field name size in bytes
      fieldSize: 1000000, // Max field value size in bytes
      fields: 10, // Max number of non-file fields
      fileSize: 100, // For multipart forms, the max file size
      files: 1, // Max number of file fields
      headerPairs: 2000 // Max number of header key=>value pairs
    }
  */
});
server.register(playlistRoute);
server.register(tvGuideRoute);
server.register(tvGuideRoute);

const start = async () => {
  try {
    await server.listen(3000);
  } catch (err) {
    console.log(err);
    server.log.error(err);
    process.exit(1);
  }
};

/*
process.on("uncaughtException", error => {
  console.error(error);
});
process.on("unhandledRejection", error => {
  console.error(error);
});

*/
start();
