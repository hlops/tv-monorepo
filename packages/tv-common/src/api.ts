interface Dictionary<T> {
  [index: string]: T;
}

export enum ActionType {
  PLAYLIST = 'playlist',
  GUIDE = 'guide'
}

export interface Action {
  id: string;
  title: string;
  cron?: string;
  history: string[];
  type: string;
  url: string;
}

export enum ChannelOptions {
  group = "group-title"
}

export interface Channel {
  title: string;
  url: string;
  options: Dictionary<string>;
}
