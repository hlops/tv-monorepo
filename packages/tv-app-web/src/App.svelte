<script>
  import MainTabs from "./MainTabs.svelte";
</script>

<main>
  <MainTabs />
</main>
