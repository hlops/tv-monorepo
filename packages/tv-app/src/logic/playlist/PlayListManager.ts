import keyFileStorage from "key-file-storage";
import { KeyFileStorage } from "key-file-storage/dist/src/key-file-storage";
import _ from "lodash";
import { QueryParams } from "../common/QueryParams";
import { Channel, ChannelHelper } from "./Channel";
import { M3uParser } from "./m3u/M3uParser";

const STORAGE_PREFIX = "channels/";

export class PlayListManager {
  constructor(
    private readonly storage: KeyFileStorage = keyFileStorage("node_data", true)
  ) {
    this.storage[STORAGE_PREFIX];
  }

  public import(stream: NodeJS.ReadableStream): Promise<void> {
    return new M3uParser().parse(stream).then(channels => {
      return Promise.all(
        _.map(channels, channel =>
          this.storage(
            `${STORAGE_PREFIX}${ChannelHelper.getHash(channel.url)}`,
            channel
          )
        )
      ).then();
    });
  }

  public find(params: QueryParams): Promise<Channel[]> {
    const first = params.page * params.pageSize;
    return this.storage(STORAGE_PREFIX)
      .then(ids =>
        Promise.all(
          _.map(ids, id =>
            this.storage(id).then(channel =>
              ChannelHelper.find(channel, params.query)
            )
          )
        )
      )
      .then(channels =>
        _.compact(channels || []).slice(first, first + params.pageSize)
      );
  }
}
