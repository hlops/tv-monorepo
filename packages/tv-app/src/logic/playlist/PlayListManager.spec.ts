import fs from 'fs';
import keyFileStorage from 'key-file-storage';
import { KeyFileStorage } from 'key-file-storage/dist/src/key-file-storage';
import { PlayListManager } from './PlayListManager';

let storage: KeyFileStorage;

describe('PlayListManager', () => {
	beforeAll(() => {
		storage = keyFileStorage('node_data/test', true);
	});

    afterAll(() => {
        delete storage['*'];
    });

    beforeEach(() => {
		delete storage['*'];
	});

    it('import() can successfully import channels', done => {
		new PlayListManager(storage)
			.import(fs.createReadStream('./samples/playlist.m3u', 'utf8'))
			.then(() => {
				expect(storage['channels/']).toMatchSnapshot();
				done();
			});
	});
});
