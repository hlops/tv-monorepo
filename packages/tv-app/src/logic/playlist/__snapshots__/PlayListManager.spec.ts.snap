// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayListManager import() can successfully import channels 1`] = `
Array [
  "channels/6ca4bf5c5e06368afefb92d12ee31c31",
  "channels/cdcfbc645b06eedaecab14aac7ed4ffd",
  "channels/c9c7b5de237a193a4645b4d5612237b8",
  "channels/dd15a6937822cd14240f6018a8ab96b4",
  "channels/955eb9104e11fa92fbbfb06bf7fc0e3a",
  "channels/c8555c15a805486a542ba162222e2e52",
  "channels/583f085d67c5aa03c2f77a86bffe9e87",
  "channels/1c73e03fed6c670e531441866c0da873",
  "channels/62fb46ec31c734d54c0d03e22c942a50",
  "channels/21d042149b078e70c9063f644bedf84a",
  "channels/a3cfe1409c5fc8d7412747c041d7b8fd",
  "channels/1b16af9e7ae43157cf8941180b0966d0",
  "channels/50b2bdfcaaf2bccd549273c2614dca72",
  "channels/b6ce5c5d1e0d2610d4a7adc20630f3ce",
  "channels/ff685e55a086610ff512e72319163e65",
  "channels/fe8aab1c21f7f14c6149e9844bb4d0ba",
  "channels/3d322dced4d79a6aca3f67a66426b1c1",
  "channels/c2f1271248889269a5620eb8e7153a83",
  "channels/b601c8bf7368f00f8338c1c635dd6087",
  "channels/c8d84ca34cfce6ffe0a00a6c8852a726",
  "channels/3bef9ee6c714ee31547de06d4ae49a90",
  "channels/6504a0c1d349656c7e291839b9eb2045",
  "channels/d97c2e398a35bd50c5c2bf627ed494f0",
  "channels/f5efd070dd555e9b5c4498798710f7a2",
  "channels/aba1c833ad625a41b2a879d648d4aca1",
  "channels/a4988194e26848f51f8026d2743ff841",
  "channels/ce90ba56d1013bae3206c37758e211ec",
  "channels/cbe355d87a2d429dd8e5349e0eeda4a1",
  "channels/3ed73df83c258c0cc9c09accbe490e18",
  "channels/cfd14ea4fe602804e68ad816579b5c83",
  "channels/6329f110e2b17833c8cf772cbdccec2f",
  "channels/0d1971dcda5b2fcbd40f3f71cf2ef874",
  "channels/7b6d5b2c766fce4b59b4363e86c0d590",
  "channels/5dcf39ba359641795d487f1223247b97",
  "channels/f96529d7a966a66e7894a3440d4b00c6",
  "channels/54857f9572708263b52ba61427ce1565",
  "channels/67e10f2cefe6893dfd135bfb6b32f617",
  "channels/6ceda0f83077595b4a5a3d7e6a209a86",
  "channels/15dcad9eeaf4e0d80890cb325d82879f",
  "channels/9d414bba379524fa73020bab8a134016",
  "channels/46c1ffb54a73cfa44974766c3326716c",
  "channels/b5f49a9083103b2a6c22c2f664610aa9",
  "channels/2221f64653a6191bfa7f1d2f0b0b044d",
  "channels/259c0eef2be31374154f5ae381e52d8e",
  "channels/c0eedd28f4f38050209e40e249765183",
  "channels/c0b851e6abf650afdaa476c503a8c82d",
  "channels/9a71d7dfd0e067d8e6fb8e085e53ff23",
  "channels/9e533d0cf898c77f2819920ad8986f04",
  "channels/85073551efe4e3ddb1fbdd6291a15c70",
  "channels/9e992b9ccaf44fdb20d314070aa6062e",
  "channels/c8e8f209d3c865cdd231927017d5e051",
  "channels/898e26c34bc865561b1b1053da358b13",
  "channels/e8c8873c0fc72106c7fdfa0491ab5b38",
  "channels/d41650b29bc236951b3a28d5acc25e7c",
  "channels/30c782355558adb227fd16ff0a6a03f8",
  "channels/92c6511ed9ec2d7610a86456c602e853",
  "channels/c331afa4457c31dea2d2b7d98bad6f50",
  "channels/30690390b8bbbd212b3c6f8c589b61a6",
  "channels/9dd0094fcf288fa3e1dee6cc6ea7348f",
  "channels/85e4427b2b121e2e466036b47111c517",
  "channels/19e2e7a029abedcc5c0ed4fca2acb7bb",
  "channels/b27e59ddbb5b00196219129e11dc8c6e",
  "channels/825cdf846ffccd3b9199ab720b5a0693",
  "channels/fa33c9a6af0702272fc9a0ed9945b348",
  "channels/d8cd95dee6aeb3f1304adb98ce55a18f",
  "channels/ff0bd3acbc939cecb47166d0547b570c",
  "channels/6ec5b0da3477b43059b80b192a338abf",
  "channels/b98f1f4039194f1ce1d75f8d39048ca0",
  "channels/524370f59908a4d9059ddf4b7485ed74",
  "channels/1fb147d4f0d52f8439d1568da5a9a77b",
  "channels/72dbe1e937dbc0818b0ffb457632bddf",
  "channels/895db7a5ed06d9d02c9b8d042410b9c4",
  "channels/3052a0101e0f3692ffab78106a65468c",
  "channels/185c9785feb489f570e948ae45aac130",
  "channels/4e453ecc5678a3397cf036ce61dac933",
  "channels/f12ad30db2a521d7e0ff6ebf69225531",
  "channels/cccc9f19192c92e85f297b776b6f9db8",
  "channels/b16f12dc5c5b98fcd8c20e440279b6b7",
  "channels/6dbee6d95e657f1d6022c812b3ca4cfd",
  "channels/107877bb4b6da32fe10c728172d32222",
  "channels/110db01de22fdd0f7bc3b3bb8d27fb4a",
  "channels/8f4ef379dde2405b6cdbd0f9df75e79a",
  "channels/12db2071154332f3b337193e46daa0d2",
  "channels/f5a62170b99e1893ee2cdee72ec3537f",
  "channels/01caba6f9e9891e279ff365d43715724",
  "channels/f27f450aba21814a962f4364fa50daff",
  "channels/7c8983de0e6c2e2aa1a4fed6b824ea56",
  "channels/f9c98b2a6d7037a41b21bf6334fe10a3",
  "channels/f1679fb5c0d20f2fd086cb47e4a4ce2e",
  "channels/5f1e9a5af28076b869135aabe5f716f3",
  "channels/228cd865e2296bbc739f6e1a07862e08",
  "channels/2ce1cd94a339366ca128ced21d4d610c",
  "channels/72ea6b4fd4cfffddadc9f54236c92f6f",
  "channels/8249c2991ef4efb69bcffe6cfa334a55",
  "channels/7a0c2c209039328302d31d4cb6badc17",
  "channels/3f579f1bb09e9a8d19beb6444b30658d",
  "channels/07550d715acb93f0362c0c1f598bf0ec",
  "channels/baa03dab49832ff8d6d549defc2d40ce",
  "channels/c33de0f84c7485001c79f754e5458db8",
  "channels/25923914d87ead7360cb43c43b7fdda0",
  "channels/5a2d09af6e70b6ef60d4f1b4f9cc495f",
  "channels/c25a70bb21fbdca57699b302d770d2ab",
  "channels/d44d0bd81345a6e60dc500d58bd47423",
  "channels/e46476a94c2f7cf03719a25ca3fb5e39",
  "channels/0dd1f09520de6f3e7e1d5276be23906c",
  "channels/fbc26a674cab503e229572335c5c3f9e",
  "channels/3f3f907b3f3bdad0a667cd50c1a4c69c",
  "channels/41e96d57b9ab25c1bad4441c93492575",
  "channels/c6b6cf1d166bf95dc473875cbb0d67e7",
  "channels/f9cb5c9e561ba9a5922e212f4dd4e99c",
  "channels/7989e77dc41bcbbd3cb1f1952e2c42fa",
  "channels/30536933fbaaf16c01aa0a825f79045e",
  "channels/384c859b3d858a846c63677a9b1ff477",
  "channels/42ca68630cc6bb309bf542f9738ec1db",
  "channels/bce936a0f01e2b4f0d604ca8d8e14554",
  "channels/a88f588a2546fb6c33aae9dda8717901",
  "channels/a7bf1b1957314091802c2efbd78a3a74",
  "channels/18f3266fa33156904e9c68673305252a",
  "channels/e4ddb6de35006b5ca574595fd383bc21",
  "channels/70e80f5fb860acc93070b55f04d45e74",
  "channels/505da93e87a4f82c9faada4c324f1d55",
  "channels/1ae4771b51a8e2666338763bfed360b7",
  "channels/be29b0be0967c3b1f2f6d4919ee7b3f2",
]
`;
